keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,436 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+The [Topology Aware Hints feature] provides a way of signaling to
+kube-proxy that it should optimize service delivery on a
+per-availability zone basis, but does not provide any way to optimize
+on any other basis.
+
+This KEP adds a new topology hint, to tell kube-proxy that a Service
+is expected to have an endpoint on every node most of the time, and so
+it should optimize traffic by always routing to a local endpoint _when
+one is available_, but using other topology hints or ordinary
+`Cluster` traffic policy when not.
+
+[Topology Aware Hints feature]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2433-topology-aware-hints
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### User Stories
+
+#### Fast/Reliable (but Non-Source-IP-Preserving) LoadBalancer Services
+
+As an application developer, I want incoming service traffic from
+outside the cluster to be delivered to my pods as efficiently as
+possible (as with `Local` traffic policy), but I don't care if the
+source IP is preserved, and I _do_ care about maximizing reliability
+when pods restart or move between nodes (which can cause hiccups with
+`externalTrafficPolicy: Local` in some cases).
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
```

Comment:
I think we likely need more nuanced subsetting if we were going to proceed with this approach.

Let's say we have a 2 zone cluster that has 3 nodes per zone. In each zone, the nodes have the following allocation:

A: 0 endpoints
B: 1 endpoint
C: 2 endpoints

A naive implementation would mean that traffic originating on A would be split evenly between endpoints on nodes B and C. If traffic was proportionally distributed from each node, that would mean the following distribution:

Endpoint B1 gets 100% of traffic from node B + 33% of traffic from node A
Endpoints C1 and C2 each get 50% of traffic from node A + 33% of traffic from node A

That distribution is not that bad - a range of 83% - 133% expected traffic, but I think it could be _much_ worse depending on how endpoints were deployed.

In the contrived scenario above, I think we'd actually want an allocation like this:

Endpoint B1 gets 100% of traffic from node B
Endpoints C1 and C2 each get 50% of traffic from node A + 50% of traffic from node A

That kind of per endpoint distribution sounds _very_ complicated and potentially not worth it, just want to be sure we properly explore the risks of uneven distribution and any potential ways we could mitigate that.

Would it make sense to rethink hints so that we don't assign hints to endpoints that will be used exclusively for local traffic with this approach?

Reply:
> In each zone, the nodes have the following allocation:
> 
> A: 0 endpoints
> B: 1 endpoint
> C: 2 endpoints

This feature is primarily intended for services where you expect that a local connection will always be possible. (ie, either node that has pods that will call the service also has an endpoint for the service). So in this case, I would say that if you expect that there are pods on Node A that will need to use the service, then the service should not use node-level topology. (ie, either you don't mark the service as using node-level topology, or we don't autodetect the service as getting node-level topology).

I guess I should be more explicit about the fact that this feature is only intended to apply to a very small percentage of services.